src/System.Net.Http/src/System/Net/Http/HPack/HPackDecoder.cs

```diff
@@ -0,0 +1,407 @@
+                            }
+                            else
+                            {
+                                _state = State.HeaderNameIndex;
+                            }
+                        }
+                        else if ((b & LiteralHeaderFieldWithoutIndexingMask) == LiteralHeaderFieldWithoutIndexingRepresentation)
+                        {
+                            _index = false;
+                            int val = b & ~LiteralHeaderFieldWithoutIndexingMask;
+
+                            if (val == 0)
+                            {
+                                _state = State.HeaderNameLength;
+                            }
+                            else if (_integerDecoder.StartDecode((byte)val, LiteralHeaderFieldWithoutIndexingPrefix))
+                            {
+                                OnIndexedHeaderName(_integerDecoder.Value);
+                            }
+                            else
+                            {
+                                _state = State.HeaderNameIndex;
+                            }
+                        }
+                        else if ((b & LiteralHeaderFieldNeverIndexedMask) == LiteralHeaderFieldNeverIndexedRepresentation)
+                        {
+                            _index = false;
+                            int val = b & ~LiteralHeaderFieldNeverIndexedMask;
+
+                            if (val == 0)
+                            {
+                                _state = State.HeaderNameLength;
+                            }
+                            else if (_integerDecoder.StartDecode((byte)val, LiteralHeaderFieldNeverIndexedPrefix))
+                            {
+                                OnIndexedHeaderName(_integerDecoder.Value);
+                            }
+                            else
+                            {
+                                _state = State.HeaderNameIndex;
+                            }
+                        }
+                        else if ((b & DynamicTableSizeUpdateMask) == DynamicTableSizeUpdateRepresentation)
+                        {
+                            if (_integerDecoder.StartDecode((byte)(b & ~DynamicTableSizeUpdateMask), DynamicTableSizeUpdatePrefix))
+                            {
+                                // TODO: validate that it's less than what's defined via SETTINGS
+                                _dynamicTable.Resize(_integerDecoder.Value);
+                            }
+                            else
+                            {
+                                _state = State.DynamicTableSizeUpdate;
+                            }
+                        }
+                        else
+                        {
+                            // Can't happen
+                            Debug.Fail("Unreachable code");
+                            throw new InternalException();
+                        }
```

Comment:
If it can't happen, then you could remove the whole else block and change the last else if above to just be an else, asserting the condition that's currently the condition of the else if.